sql/core/src/main/java/org/apache/spark/sql/execution/datasources/orc/OrcColumnarBatchReader.java

```diff
@@ -0,0 +1,509 @@
+   * Please note that `initBatch` is needed to be called after this.
+   */
+  @Override
+  public void initialize(InputSplit inputSplit, TaskAttemptContext taskAttemptContext)
+      throws IOException, InterruptedException {
+    FileSplit fileSplit = (FileSplit)inputSplit;
+    Configuration conf = taskAttemptContext.getConfiguration();
+    reader = OrcFile.createReader(
+      fileSplit.getPath(),
+      OrcFile.readerOptions(conf)
+        .maxLength(OrcConf.MAX_FILE_LENGTH.getLong(conf))
+        .filesystem(fileSplit.getPath().getFileSystem(conf)));
+
+    Reader.Options options =
+      OrcInputFormat.buildOptions(conf, reader, fileSplit.getStart(), fileSplit.getLength());
+    recordReader = reader.rows(options);
+    totalRowCount = reader.getNumberOfRows();
+  }
+
+  /**
+   * Initialize columnar batch by setting required schema and partition information.
+   * With this information, this creates ColumnarBatch with the full schema.
+   */
+  public void initBatch(
+      TypeDescription orcSchema,
+      int[] requestedColIds,
+      StructField[] requiredFields,
+      StructType partitionSchema,
+      InternalRow partitionValues) {
+    batch = orcSchema.createRowBatch(DEFAULT_SIZE);
+    assert(!batch.selectedInUse); // `selectedInUse` should be initialized with `false`.
+
+    this.requiredFields = requiredFields;
+    this.requestedColIds = requestedColIds;
+    assert(requiredFields.length == requestedColIds.length);
+
+    StructType resultSchema = new StructType(requiredFields);
+    for (StructField f : partitionSchema.fields()) {
+      resultSchema = resultSchema.add(f);
+    }
+
+    int capacity = DEFAULT_SIZE;
+    if (MEMORY_MODE == MemoryMode.OFF_HEAP) {
+      columnVectors = OffHeapColumnVector.allocateColumns(capacity, resultSchema);
+    } else {
+      columnVectors = OnHeapColumnVector.allocateColumns(capacity, resultSchema);
+    }
+    columnarBatch = new ColumnarBatch(resultSchema, columnVectors, capacity);
+
+    if (partitionValues.numFields() > 0) {
+      int partitionIdx = requiredFields.length;
+      for (int i = 0; i < partitionValues.numFields(); i++) {
+        ColumnVectorUtils.populate(columnVectors[i + partitionIdx], partitionValues, i);
+        columnVectors[i + partitionIdx].setIsConstant();
+      }
+    }
+
+    // Initialize the missing columns once.
+    for (int i = 0; i < requiredFields.length; i++) {
+      if (requestedColIds[i] == -1) {
+        columnVectors[i].putNulls(0, columnarBatch.capacity());
+        columnVectors[i].setIsConstant();
+      }
+    }
+  }
+
+  /**
+   * Return true if there exists more data in the next batch. If exists, prepare the next batch
+   * by copying from ORC VectorizedRowBatch columns to Spark ColumnarBatch columns.
+   */
+  private boolean nextBatch() throws IOException {
+    if (rowsReturned >= totalRowCount) {
+      return false;
+    }
+
+    recordReader.nextBatch(batch);
+    int batchSize = batch.size;
+    if (batchSize == 0) {
```

Comment:
Sorry to be nitpicking, but why would PPD affect this? Do you mean PPD would make the `totalRowCount` incorrect?

Reply:
Yes. Right. At the beginning, we are using a total row number which is the real total row count of the file. PPD can be compute during actually filling batch.

Reply:
will we output less data? Assume one batch has 0 rows after PPD, but the ORC file has more batches to return, what should we do?

Also we may report wrong progress, since the `totalRowCount` is the number of rows in the ORC file, while `rowsReturned` is the number of rows returned after PPD.

Reply:
Here, batch is a unit of ORC reader. A batch is 0 means it's at the end. It will not occur when there is more remaining rows.

Reply:
I'm a little confused, do you mean the ORC reader will skip empty batch(due to PPD) until it's the end?

BTW does Hive report progress correctly?

Reply:
Maybe, I explained in a vague manner. Sorry for that. Let me put in this way.

ORC reader is the one who making batches. ORC reads the raw data and makes batches with filtered rows only. So, you can ignore the original rows. From now, please assume that you have only qualified rows. ORC will privide users a batch with fully packed in most cases. The last batch will be partially packed (batchSize < the max batch size) or empty.

Reply:
I see. Parquet is fine as parquet reader in Spark only do row-group PPD, not record level PPD. I'm OK to leave it since progress reporting is not very important, but it would be great to take a look at how hive solves this issue.

Reply:
For the progress reporting, I checked it again and we can delegate it to the record reader. I'll update like that.

- Hive: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/io/orc/OrcNewInputFormat.java#L112
- ORC: https://github.com/apache/orc/blob/master/java/mapreduce/src/java/org/apache/orc/mapreduce/OrcMapreduceRecordReader.java#L126